### PR TITLE
TASK: Export event cleanup and error handling for catchupd events

### PR DESCRIPTION
**Pull request by mhsdesign**
We fixed the root source for the frequent problem of workspace rebase failed events having a too big payload https://github.com/neos/neos-development-collection/issues/4545, which might be clamped when dumping the events table and importing it elsewhere where it cannot be decoded.

But i think enriching the error with the sequence number makes sense, as the event id might not really be known to developers.

**Upgrade instructions**

**Review instructions**

**Checklist**

- [ ] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

### CODE REVIEW

File: Neos.ContentRepository.Export/src/Event/ValueObject/ExportedEvent.php [hunk 1]
```diff
@@ -7,39 +7,39 @@
...
+     * This exporter will enforce this as well (by using array as payload type).
+     *
      * @param array<mixed> $payload
      * @param array<mixed> $metadata
      */
     public function __construct(
         public readonly string $identifier,
         public readonly string $type,
-        public readonly array $payload, // TODO: string
+        public readonly array $payload,
```
Comment (mhsdesign) [inline Neos.ContentRepository.Export/src/Event/ValueObject/ExportedEvent.php]: fyi i tried to solve the todo https://github.com/neos/neos-development-collection/pull/4917 but its senseless ^^

File: Neos.ContentRepository.Export/src/Event/ValueObject/ExportedEvent.php [hunk 2]
```diff
@@ -80,12 +80,8 @@ public function toJson(): string
         }
     }
 
-    /**
-     * @return array{identifier: string, type: string, payload: array<mixed>, metadata?: ?array<mixed>, streamName?: string, version?: int, sequenceNumber?: int, recordedAt?: string}
-     */
-    public function jsonSerialize(): array
+    public function jsonSerialize(): mixed
```
Comment (bwaidelich) [inline Neos.ContentRepository.Export/src/Event/ValueObject/ExportedEvent.php:83]: This returns an array, we should not make the API less specific IMO, but I'd be fine with:

```suggestion
    /**
      *@return array<string, mixed>
      */
    public function jsonSerialize(): array
```

File: Neos.ContentRepository.Core/Classes/Projection/CatchUp.php [hunk 3]
```diff
@@ -104,7 +104,11 @@ public function run(EventStreamInterface $eventStream): SequenceNumber
                 if ($eventEnvelope->sequenceNumber->value <= $highestAppliedSequenceNumber->value) {
                     continue;
                 }
-                ($this->eventHandler)($eventEnvelope);
+                try {
+                    ($this->eventHandler)($eventEnvelope);
+                } catch (\Exception $e) {
```
Comment (bwaidelich) [inline Neos.ContentRepository.Core/Classes/Projection/CatchUp.php:109]:
```suggestion
                } catch (\Throwable $e) {
```

File: Neos.ContentRepository.Core/Classes/Projection/CatchUp.php [hunk 4]
```diff
@@ -104,7 +104,11 @@ public function run(EventStreamInterface $eventStream): SequenceNumber
...
+                    throw new \RuntimeException(sprintf('Exception while catching up to sequence number %d', $eventEnvelope->sequenceNumber->value), 1710707311, $e);
```
Comment (bwaidelich) [inline Neos.ContentRepository.Core/Classes/Projection/CatchUp.php:110]: Couldn't this lead to unwanted side effects when catching specific exceptions?

File: Neos.ContentRepository.Export/src/Event/ValueObject/ExportedEvent.php [hunk 5]
```diff
@@ -7,39 +7,39 @@
...
-            /** @var array<mixed> $data */
+            /** @var array{identifier: string, type: string, payload: array<mixed>, metadata: array<mixed>} $data */
             $data = json_decode($json, true, 512, JSON_THROW_ON_ERROR);
         } catch (\JsonException $e) {
             throw new \InvalidArgumentException(sprintf('Failed to decode JSON "%s": %s', $json, $e->getMessage()), 1638432979, $e);
         }
-        assert(isset($data['identifier']) && is_string($data['identifier']));
-        assert(isset($data['type']) && is_string($data['type']));
-        assert(isset($data['payload']) && is_array($data['payload']));
-        assert(isset($data['metadata']) && is_array($data['metadata']));
```
Comment (bwaidelich) [inline Neos.ContentRepository.Export/src/Event/ValueObject/ExportedEvent.php:42]: Why did you remove the assertions?
I'm fine with replacing `assert()` calls by actual assertions, but we should not just remove them IMO

**Review by bwaidelich — DISMISSED** [review]
Nice improvement, just a few comments though

File: Neos.ContentRepository.Export/src/Event/ValueObject/ExportedEvent.php (hunk 2 above)
Comment (mhsdesign) [reply Neos.ContentRepository.Export/src/Event/ValueObject/ExportedEvent.php:83]: I got used to this pattern because

a. returning the actual type will make things more verbose, e.g. our phpstan will require the iterable to be specified for `arrays` and this is just obsolete annotation in this case
b. the method should never be called directly from a user as `jsonSerialize` is only for `json_encode` which is highly dynamic either way. In https://github.com/neos/neos-development-collection/pull/4156 we refactored all our `jsonSerialize` calls to actual value accessors and as we dont want to use `jsonSerialize` as php method, there is no gain in annotating it, and it will make usages of `jsonSerialize` harder as it is also not to be used.

File: Neos.ContentRepository.Core/Classes/Projection/CatchUp.php (hunk 3 above)
Comment (mhsdesign) [reply Neos.ContentRepository.Core/Classes/Projection/CatchUp.php:109]: In this case i deliberately only handle `Exceptions` as they are somewhat more planned (like the `InvalidArgumentException` from the event normaliser. `Errors` should just not happen (except occasional `TypeErrors` i guess, but event they should just bubble up all the way). If you have a better philosophy im open for it ;)

File: Neos.ContentRepository.Core/Classes/Projection/CatchUp.php (hunk 4 above)
Comment (mhsdesign) [reply Neos.ContentRepository.Core/Classes/Projection/CatchUp.php:110]: Hmm maybe we should create a specific `CatctupException` here but im not aware of any specific exceptions thrown and `catchUpProjection` is never try catched.

File: Neos.ContentRepository.Export/src/Event/ValueObject/ExportedEvent.php (hunk 5 above)
Comment (mhsdesign) [reply Neos.ContentRepository.Export/src/Event/ValueObject/ExportedEvent.php:42]: This should not happen as this is the assumed shape of the array.
And with the type annotations phpstan will even scream if we double check because we annotated that this is the shape.

Also in case there is a low level error a dedicated `TypeError` will be thrown which is much more explicit than this.

File: Neos.ContentRepository.Core/Classes/Projection/CatchUp.php (hunk 4 above)
Comment (bwaidelich) [reply Neos.ContentRepository.Core/Classes/Projection/CatchUp.php:110]: I meant if a projection were to throw a specific exception we would now "mask" that

File: Neos.ContentRepository.Export/src/Event/ValueObject/ExportedEvent.php (hunk 2 above)
Comment (bwaidelich) [reply Neos.ContentRepository.Export/src/Event/ValueObject/ExportedEvent.php:83]: ok, good points! the thing is, that we keep discussing these things and changing them back and forth.
IMO this change is not related to this PR

File: Neos.ContentRepository.Core/Classes/Projection/CatchUp.php (hunk 3 above)
Comment (bwaidelich) [reply Neos.ContentRepository.Core/Classes/Projection/CatchUp.php:109]: Makes sense

File: Neos.ContentRepository.Export/src/Event/ValueObject/ExportedEvent.php (hunk 5 above)
Comment (bwaidelich) [reply Neos.ContentRepository.Export/src/Event/ValueObject/ExportedEvent.php:42]: yes, right, I overlooked that the types are assumed right below

File: Neos.ContentRepository.Export/src/Event/ValueObject/ExportedEvent.php (hunk 2 above)
Comment (mhsdesign) [reply Neos.ContentRepository.Export/src/Event/ValueObject/ExportedEvent.php:83]: okay yes that makes sense i will try to create a canvas where we can discuss new code styles and put agreed upon practices in a separate spot?

I changed the code as i wanted to get rid of the phpstan ignore, originally this change is of course picked together a bit and contains / solves separate concerns but i didnt want to just trash my local stash, thus i created a pr which kinda makes sense hopefully: Event (export) related cleanup

File: Neos.ContentRepository.Core/Classes/Projection/CatchUp.php (hunk 4 above)
Comment (mhsdesign) [reply Neos.ContentRepository.Core/Classes/Projection/CatchUp.php:110]: ha 

>  projection were to throw a specific exception

this is a trick question ;) a projection should never throw :D

Comment (bwaidelich) [reply Neos.ContentRepository.Core/Classes/Projection/CatchUp.php:110]: You passed the test 😀

**Review by bwaidelich — APPROVED** [review]
+1 by reading
